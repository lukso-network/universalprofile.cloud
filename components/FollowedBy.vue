<script setup lang="ts">
type Props = {
  followerAddresses?: Address[]
  followingAddresses?: Address[]
}

const props = defineProps<Props>()
const { formatMessage } = useIntl()

const followers = computed(() => {
  return (
    props.followerAddresses?.filter(address =>
      props.followingAddresses?.includes(address)
    ) || []
  )
})

const restOfFollowers = computed(() => {
  return followers.value.slice(1, 3)
})

const addressesForFollowerNames = computed(() => {
  return followers.value.slice(0, 2)
})

const othersCount = computed(() => {
  return (
    followers.value.length >= 2 ? followers.value.length - 2 : 0
  ).toString()
})

const profileNames = (profiles?: Profile[]) =>
  profiles
    ?.map((profile: Profile) => {
      return `<lukso-username
        name='${profile?.name || formatMessage('profile_default_name')}'
        address='${profile.address}'
        size='small'
        address-color='neutral-20'
        name-color='neutral-20'
        ${profile?.name ? '' : 'hide-prefix'}
      ></lukso-username>`
    })
    .join(', ') || ''
</script>

<template>
  <div v-if="followers.length > 0" class="flex items-center">
    <div class="flex space-x-[-14px]">
      <LoaderProfile
        v-for="(followerAddress, index) in restOfFollowers || []"
        :profile-address="followerAddress"
        :key="index"
      >
        <template #default="{ profile, profileAvatar }">
          <lukso-profile
            size="x-small"
            :profile-url="profileAvatar?.url"
            :profile-address="profile?.address"
          ></lukso-profile>
        </template>
      </LoaderProfile>
      <LoaderProfile v-if="followers[0]" :profile-address="followers[0]">
        <template #default="{ profile, profileAvatar }">
          <lukso-profile
            size="x-small"
            :profile-url="profileAvatar?.url"
            :profile-address="profile?.address"
          ></lukso-profile>
        </template>
      </LoaderProfile>
    </div>
    <div class="paragraph-inter-12-medium pl-2">
      <LoaderProfiles :profile-addresses="addressesForFollowerNames">
        <template #default="{ profiles, isLoading }">
          <AppPlaceholderLine v-if="isLoading" class="h-[15px] w-[200px]" />
          <lukso-sanitize
            v-else
            :html-content="
              formatMessage('followed_by_text', {
                names: profileNames(profiles),
                othersCount,
              })
            "
          >
          </lukso-sanitize>
        </template>
      </LoaderProfiles>
    </div>
  </div>
</template>
