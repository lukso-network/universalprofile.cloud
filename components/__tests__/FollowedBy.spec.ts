import { renderSuspended } from '@nuxt/test-utils/runtime'
import { beforeEach, describe, expect, it, vi } from 'vitest'

import FollowedBy from '../FollowedBy.vue'

const profilesMock = vi.fn()

vi.mock('/composables/useProfiles', () => ({
  useProfiles: () => profilesMock,
}))

describe('FollowedBy', () => {
  beforeEach(() => {
    useIntl().setupIntl(defaultConfig)
    profilesMock.mockReturnValue({
      profiles: [],
      isLoading: false,
    })
  })

  it('should render empty state', async () => {
    const component = await renderSuspended(FollowedBy)

    expect(component.html()).toMatchSnapshot()
  })

  it('should render for 1 follower', async () => {
    profilesMock.mockReturnValue(
      toRef({
        profiles: [
          {
            address: '0x1',
            name: 'john',
          },
        ],
        isLoading: false,
      })
    )
    const component = await renderSuspended(FollowedBy, {
      props: {
        followerAddresses: ['0x1'],
        followingAddresses: ['0x1'],
      },
    })

    expect(component.html()).toMatchSnapshot()
  })

  it('should render for 2 followers', async () => {
    profilesMock.mockReturnValue(
      toRef({
        profiles: [
          {
            address: '0x1',
            name: 'john',
          },
          {
            address: '0x2',
            name: 'jane',
          },
        ],
        isLoading: false,
      })
    )
    const component = await renderSuspended(FollowedBy, {
      props: {
        followerAddresses: ['0x1', '0x2'],
        followingAddresses: ['0x1', '0x2'],
      },
    })

    expect(component.html()).toMatchSnapshot()
  })

  it('should render for 3 followers', async () => {
    profilesMock.mockReturnValue(
      toRef({
        profiles: [
          {
            address: '0x1',
            name: 'john',
          },
          {
            address: '0x2',
          },
        ],
        isLoading: false,
      })
    )
    const component = await renderSuspended(FollowedBy, {
      props: {
        followerAddresses: ['0x1', '0x2', '0x3'],
        followingAddresses: ['0x1', '0x2', '0x3'],
      },
    })

    expect(component.html()).toMatchSnapshot()
  })

  it('should render for 4 followers', async () => {
    profilesMock.mockReturnValue(
      toRef({
        profiles: [
          {
            address: '0x1',
            name: 'john',
          },
          {
            address: '0x2',
            name: 'jane',
          },
        ],
        isLoading: false,
      })
    )
    const component = await renderSuspended(FollowedBy, {
      props: {
        followerAddresses: ['0x1', '0x2', '0x3', '0x4'],
        followingAddresses: ['0x1', '0x2', '0x3', '0x4'],
      },
    })

    expect(component.html()).toMatchSnapshot()
  })

  it('should render for 5 followers', async () => {
    profilesMock.mockReturnValue(
      toRef({
        profiles: [
          {
            address: '0x1',
            name: 'john',
          },
          {
            address: '0x2',
            name: 'jane',
          },
        ],
        isLoading: false,
      })
    )
    const component = await renderSuspended(FollowedBy, {
      props: {
        followerAddresses: ['0x1', '0x2', '0x3', '0x4', '0x5'],
        followingAddresses: ['0x1', '0x2', '0x3', '0x4', '0x5'],
      },
    })

    expect(component.html()).toMatchSnapshot()
  })
})
