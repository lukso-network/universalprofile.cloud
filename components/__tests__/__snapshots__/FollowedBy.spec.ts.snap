// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FollowedBy > should render empty state 1`] = `
"<div id="test-wrapper">
  <!--v-if-->
</div>"
`;

exports[`FollowedBy > should render for 1 follower 1`] = `
"<div id="test-wrapper">
  <div class="flex items-center">
    <div class="flex space-x-[-14px]">
      <lukso-profile size="x-small" profile-address="0x1"></lukso-profile>
    </div>
    <div class="paragraph-inter-12-medium pl-2">
      <lukso-sanitize html-content="Followed by <strong>@john#1</strong> you follow"></lukso-sanitize>
    </div>
  </div>
</div>"
`;

exports[`FollowedBy > should render for 2 followers 1`] = `
"<div id="test-wrapper">
  <div class="flex items-center">
    <div class="flex space-x-[-14px]">
      <lukso-profile size="x-small" profile-address="0x2"></lukso-profile>
      <lukso-profile size="x-small" profile-address="0x1"></lukso-profile>
    </div>
    <div class="paragraph-inter-12-medium pl-2">
      <lukso-sanitize html-content="Followed by <strong>@john#1</strong>, <strong>@jane#2</strong> you follow"></lukso-sanitize>
    </div>
  </div>
</div>"
`;

exports[`FollowedBy > should render for 3 followers 1`] = `
"<div id="test-wrapper">
  <div class="flex items-center">
    <div class="flex space-x-[-14px]">
      <lukso-profile size="x-small" profile-address="0x2"></lukso-profile>
      <lukso-profile size="x-small" profile-address="0x3"></lukso-profile>
      <lukso-profile size="x-small" profile-address="0x1"></lukso-profile>
    </div>
    <div class="paragraph-inter-12-medium pl-2">
      <lukso-sanitize html-content="Followed by <strong>@john#1</strong>, <strong>0x2</strong> and 1 other you follow"></lukso-sanitize>
    </div>
  </div>
</div>"
`;

exports[`FollowedBy > should render for 4 followers 1`] = `
"<div id="test-wrapper">
  <div class="flex items-center">
    <div class="flex space-x-[-14px]">
      <lukso-profile size="x-small" profile-address="0x2"></lukso-profile>
      <lukso-profile size="x-small" profile-address="0x3"></lukso-profile>
      <lukso-profile size="x-small" profile-address="0x1"></lukso-profile>
    </div>
    <div class="paragraph-inter-12-medium pl-2">
      <lukso-sanitize html-content="Followed by <strong>@john#1</strong>, <strong>@jane#2</strong> and 2 others you follow"></lukso-sanitize>
    </div>
  </div>
</div>"
`;

exports[`FollowedBy > should render for 5 followers 1`] = `
"<div id="test-wrapper">
  <div class="flex items-center">
    <div class="flex space-x-[-14px]">
      <lukso-profile size="x-small" profile-address="0x2"></lukso-profile>
      <lukso-profile size="x-small" profile-address="0x3"></lukso-profile>
      <lukso-profile size="x-small" profile-address="0x1"></lukso-profile>
    </div>
    <div class="paragraph-inter-12-medium pl-2">
      <lukso-sanitize html-content="Followed by <strong>@john#1</strong>, <strong>@jane#2</strong> and 3 others you follow"></lukso-sanitize>
    </div>
  </div>
</div>"
`;
